transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -343,168 +356,198 @@ private String generateName(ChannelHandler handler) {
             cache.put(handlerType, name);
         }
 
-        // It's not very likely for a user to put more than one handler of the same type, but make sure to avoid
-        // any name conflicts.  Note that we don't cache the names generated here.
-        if (context0(name) != null) {
-            String baseName = name.substring(0, name.length() - 1); // Strip the trailing '0'.
-            for (int i = 1;; i ++) {
-                String newName = baseName + i;
-                if (context0(newName) == null) {
-                    name = newName;
-                    break;
+        synchronized (handlers) {
+            // It's not very likely for a user to put more than one handler of the same type, but make sure to avoid
+            // any name conflicts.  Note that we don't cache the names generated here.
+            if (context(name) != null) {
+                String baseName = name.substring(0, name.length() - 1); // Strip the trailing '0'.
+                for (int i = 1;; i ++) {
+                    String newName = baseName + i;
+                    if (context(newName) == null) {
+                        name = newName;
+                        break;
+                    }
                 }
             }
         }
+
         return name;
     }
 
     private static String generateName0(Class<?> handlerType) {
         return StringUtil.simpleClassName(handlerType) + "#0";
     }
 
+    private void remove(AbstractChannelHandlerContext ctx) {
+        assert ctx != null;
+        EventExecutor executor = executor();
+        if (executor.inEventLoop()) {
+            remove0(ctx);
+        } else {
+            executor.execute(() -> remove0(ctx));
+        }
+    }
+
+    private int findCtxIdx(Predicate<AbstractChannelHandlerContext> predicate) {
```

Comment:
maybe have an equivalent `private AbstractChannelHandlerContext findCtx(...)` method? Quite a few of the usages only use the index to re-lookup the context

Reply:
done